fix(schema): use scule types for `runtimeConfig` type hints

luc122c opened the pull request:
<!---
☝️ PR title should follow conventional commits (https://conventionalcommits.org)

Please carefully read the contribution docs before creating a pull request
 👉 https://nuxt.com/docs/community/contribution
-->

### 🔗 Linked issue

Resolves #23246

<!-- Please ensure there is an open issue and mention its number as #123 -->

### ❓ Type of change

<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->

- [ ] 📖 Documentation (updates to the documentation, readme or JSdoc annotations)
- [x] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

In the original code, if the first character was a digit, the code was adding an underscore before the character and then converting it to uppercase. This was causing the additional underscores when the input string was already in upper snake case.

[TypeScript Playground link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAogHsATgQwMbAKpkogwgCwFcA7AawB4AVKCBCYgEwGcomkBLYgcwD4oBeKNVrB6zKAAMAJAG9OAMwiIoAQQC+shUqgAhNRIBQUKAH5VNOoxZYcqZEwjkVPI8dO6Loq5Nk2ldh3I2RE5eDTliRWUAJQg2fVc3dx1PMRZpCKioXHDY+MMkpLMM9Vkc2XgkNExsJQISCjzgHgTCtwAuKGIIADclRI7VAc6AchHXTu6+xAMDUEgoAGUAG3ZUCAAZe2AqVO9g0L5BYUtxDK1lUsztPQLkvbPNSO1y6+UAEVa3M3eHljGBsYzCpAVBOiVwq8VmtNttyO8WndjKNxsiur1+nNwNA-IhFsRkKQILh7I4Tl5xAduAAaJbAZCiP5QEZsZCIYAjKAAH2ZywA9gB3JScnkjQi1RCcwQs+nskZHIRM87PS7hC66L7Apm4gKOZwDLUiNI+GQARm5UAATBaAMwWgAsFoArBaAGwWgDsFoAHBaAJwWgAMXyKJquuPxhOJpPIOlpI35QsliNBnUW9MZRu8CcFwotYolqLa9yzjxk0PWWzY5EqKHQuPqZFjPBaTyyTRDxbMFdh1dr1QbRCbOhbTIBxcKxVk4YlkaJJMCceZrLlKYnSXB0-CParO379YljYoI9bMgjBPnMaa8ZXHLX686KVL1gluuboLaxQA+lvZLcP4U4I-jIM44HO0aLvGibCvebRphm0DPsusocgBQJhuE55RgujhLjmSbyp2cHotMqZ0gyiGnP8t4iryuaSgBU5nq+MbOJhs4XhBuFQfRhFIoBkjATqrGnlhl6QXRBH3p0wmBJQLhYgsADy0QAJIAOKqQAcioGwCFAYlceQIwqAACqZX46QAsjAizygYAD0DlJAAeiYinQJQtnUIIhk4cZZmmfZTmue58yed51q+Rx2ExiMyDYMFzluG5HlCN5drRWBnH+SMXB8nyDDqSgPQgEloVpV5iyUI6WVKOBuVICAlD4OwTBKd0WkiOVKUmEAA)


This PR also adds a couple of tests to catch values already in upper snake case (as in the original issue).

### 📝 Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- If your change requires a documentation PR, please link it appropriately -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

luc122c committed:
fix: UpperSnakeCase type not correctly handling strings already in upper snake case

luc122c committed:
ci: add tests for values already in upper snake case

luc122c committed:
Merge branch 'main' into upper-snake-case-fix

danielroe reviewed (approved):
Thank you ❤️

danielroe reviewed (changes requested):
Would you check the type fixture failures? 🙏

luc122c committed:
ci: use `assertType` syntax for type tests
See https://vitest.dev/guide/testing-types.html#reading-errors

luc122c committed:
ci: remove `only`

luc122c committed:
ci: fix TypeScript warning for tests

danielroe commented:
I believe that is how the type assertion indicates there is a mismatch. This is what the mismatch looks like:

<img width="768" alt="CleanShot 2023-10-16 at 22 55 37@2x" src="https://github.com/nuxt/nuxt/assets/28706372/42a366d1-7e8b-4c54-87e0-a72b1e42200a">

danielroe committed:
test: remove `[][0]`

danielroe committed:
Merge remote-tracking branch 'origin/main' into upper-snake-case-fix

luc122c commented:
Oh, interesting. I wasn't seeing any errors in the console output so I didn't think the tests were failing. I'll take another look.

luc122c committed:
Fix two of the tests

luc122c committed:
test: format input in camelCase

luc122c commented:
@danielroe I've changed the input for the last test that was failing. It was originally `baseAPIToken`, now `baseApiToken`, and I understand why this was struggling. Although we know that 'API' is a common acronym, TypeScript doesn't. It would be possible to add a case to check for this, however I think it may be more beneficial to correctly handle camelCase than to correctly handle this particular case? Please let me know your thoughts.

danielroe commented:
The intent is to split the string with `_` immediately before the final capital letter of a sequence of capital letters. So:

```
abcDEFghi => abc_DE_FGHI
abcDefGhi => ABC_DEF_GHI
```

luc122c committed:
Revert "test: format input in camelCase"

This reverts commit dce36555961704ae8abab2b0851ccc83572e0c68.

luc122c committed:
test: respect existing rules for conversion

luc122c commented:
> The intent is to split the string with `_` immediately before the final capital letter of a sequence of capital letters.

Ok, i've implemented [this change](https://www.typescriptlang.org/play?#code/PTAEBcE8AcFNQKrTgJwMoDsCGBrWBhLAZ1gB4AVUWAD3FgwBMjQjwUBLDAcwBpQ1wWOlVr0moAOSssKcBNAAfSQBsA9gHdYKeUokBXZFvkBeSdNkSAfKFOUadRswAGAEgDenAGZbQAQQC+7l4+AEL+TgBQIKAxAPx+Ig7iSKgAxsRkvpZRYDFxCfZizu4AjIqgAEzlAMzlACzlAKzlAGzlAOzlABzlAJzlAAzhOXmj8a5uAe4pWpi4BBmkIXwSapralsPRozEAXPyCwoWOKhpG5fqG2iM7efEhiUWgE2jK7KmwADLE4KQAorQUFhUuAZih8AALPQYHBLSybIIYbwoUAAJVgrC2uVud34bw+31Y-0BwNBV0h0NhIXhjxOEgkNxxY2eAH13AE2W4wXM8IQSEsVuY5JtItsmXl9k5OVM3K93l8fsS2KSwRSYXDNpzudheYt0axBYILCLGUz9g9jskrul+dTTeLxtLAm4wqLseLQJKndMrjyFraVmsjCaxbd9gIhPBLcwpEa5PbbuN2c7tfM+WRlqd1lYsR7PaAMLAAG5aBP5iNHUR0oU6LNGMv5CZgm2ZBFc306-0ZwNnDa5s2sn1pRZZFMdtOLTOrXs5t0SxDWxbkbIRCJQOCgAHKkGqqHquxV8SsDjcay2WniCbBFEy6+gV0RfK+C-MZsj7LMi2H4rt4f84+cFwbZ3vq4DDIm94vs8iLIqA+DOqB4E4kmkzOvB7hbkCO7knusKgSaA6FiWKCPjs+y+KReySAyVFEaWa4wPAcoEoqB5JMwAGnjYoBsU8V5Ij4t4CSiD75F+7HQR4wlwc6d4ACJIfkclQfSlG4hRYaSTK6GyviCpEnJBHzqptHFvRq7rvAqa6vyvEnJxvAHJGKk1hcQbaBcBioCYZhxlY3F2ZeMGCbJ0miaA8TPtGC5-q2amRVBExlEoVRKLUSgNEozRKG0SidEoPRKP0ShDJEEETDK1ldgKdZ9mVZFOZWEnTtmnlXDRyGQdFLx6YSvyYSqOGUhqoWwYh9VMvEzH6f1JLYagapUjS0UmXmKGVeONndr5MjCop4pesm7jTX1SpYWSC24SNWqbdVoGGrts5qZp4lPG+tofmtg6oe44UeodP2-rMnbpjVLXBvtozhocUbfjtFjPTs61jqgfqg1O7lPf9BZmSRmkVrDzWubomOI42Q5aC2pCjhT6Ag5OPbZkZmlSrTVM00DdMTgGtVY-m71kMuq4MRuADyqIAJIAOISwAcr4nzcVVoMSL4AAKassvLACyfxoFYpoAHqxCL8DkHrlCmMriyqxrBuhsbps8RbVRW7dKtYMg9vuo7lnO2g5C1G7qP0-yEhcKoqgMFLQJFpA3ujL7jH++QDTB8D3NkBIbCQOQELsEQouFrL9gJ3kScbubAfNOnXNbaQEgAEYZOrEvkKoeAYGXMSO0AA). 

`baseAPIToken` now becomes `BASE_API_TOKEN` however `APP_NAMES`, which is already correct, now becomes `A_PP_NAME_S` which was a case highlighted in the original issue.

luc122c committed:
Merge branch 'main' into upper-snake-case-fix

luc122c committed:
ci: implement check for when input is already in upper case

luc122c commented:
Ok, so I think I've got to the bottom of this. The type works fine, except for when the input is already in UPPER_SNAKE_CASE. To get around this, I implemented this check:

```ts
type UpperUpperSnakeCase<T extends string> = T extends Uppercase<T> ? T : UpperSnakeCase<T>
```

Theoretcically, this could be combined into `UpperSnakeCase`, but when I do that, it breaks other tests 😭 [See here](https://www.typescriptlang.org/play?ts=5.2.2#code/C4TwDgpgBAogHsATgQwMbAKpkogwgCwFcA7AawB4AVKCBCYgEwGcomkBLYgcwD4oBeKNVrB6zKAAMAJAG9OAMwiIoAQQC+shUqgAhNRIBQUKAH5VNOoxZYcqZEwjkVPI8dO6Loq5Nk2ldh3I2RE5eDTliRWUAJQg2fVc3dx1PMRZpCKioXHDY+MMkpLMM9Vkc2XgkNExsJQISCjzgHgTCtwAuKGIIADclRI7VAc6AchHXTu6+xAMDUEgoAGUAG3ZUCAAZe2AqVO9g0L5BYUtxDK1lUsztPQLkvbPNSO1y6+UAEVa3M3eHljGBsYzCpAVBOiVwq8VmtNttyO8WndjKNxsiur1+nNwNA-IhFsRkKQILh7I4Tl5xAduAAaJbAZCiP5QEZsZCIYAjKAAH2ZywA9gB3JScnkjQi1RCcwQs+nskZHKAGclpKC4gJkvhmaidJVM87PS7hC66L7Aplq0lOFxFcwiFUZACM3KgACZnQBmZ0AFmdAFZnQA2Z0AdmdAA5nQBOZ0ABi+Nohvgl+MJxMtOlpI35QsliNBnUW9MZdu8WcFwudYolqLa9xLjxk0PWWzY5EqKHQuPqZHIOh4LSeWSa8drZibsNb7eqXaIPb7fHr-xrte+PhkV1xKaJJMCGeZrLleZXSXBAH1ZFdxy2dlPOxLuxR52pzzJNwTt5ampmDxyj8fOiki6qhK6q9tax5ApIL5XLcoJtGeF7hG+qY7o4e5ljm8ojiedIMtAQEymyHJwauiavsm75prumbZsKf4rpMGIzPBuHFqc-w-iKvLlpKJHFEmtiWs4SEUSh6Y0TxWFIoUCHkYJgTCQJShblRaESZhf6dBagSUC4WILAA8tEACSADixkAHIqBsAjATgKmoeQIwqAACi5p5WQAsjAizygYAD0-lJAAeiY+nQJQPnUIIyEfoEzluX5gUhWF8wRVFboxaJcWOCMyDYElQVuKF4VCFFnpZfZlGOSMXB8nyDCmSgPQgIVKWlZFiyUD6lXKdVlojEgICUPg7BMAZ3QWSIbXFal2JlV1-q9Xi-XxQARqSrnGZQfJEsQM3GCVaULZQQbLQ5A0bQ4GDRBsB1QEd82dZQobnatuVgIQa2rKguB8pE7BcPdj0LM9EZvWJ8WiGwf0A0DLjJbNHVRdGEM5U5yDA3NoNRQ6Ma2bFqlOQ6LruljQA)

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

danielroe commented:
What about test_APP_NAME?

luc122c committed:
test: add additional runtime test

luc122c commented:
`test_APP_NAME` becomes `TEST__APP_NAME`. I presume we'd like it to be `TEST_APP_NAME`? I've added an additional test case.

luc122c committed:
test: add additional type check for values beginning with an underscore

luc122c commented:
Ok, I think it's done now. All the tests pass, including the new one. 👍

luc122c committed:
Merge branch 'main' into upper-snake-case-fix

luc122c committed:
Merge branch 'main' into upper-snake-case-fix

luc122c commented:
@danielroe The test failure doesn't seem to be related to this PR:
```sh
listen EADDRINUSE: address already in use 127.0.0.1:59823
```
Please can the tests be re-run?

danielroe committed:
Merge remote-tracking branch 'origin/main' into upper-snake-case-fix

danielroe committed:
test: add more type assertion tests to cover other cases

danielroe committed:
test: remove runtime tests

luc122c committed:
test: add test cases for a single character

luc122c committed:
Merge branch 'main' into upper-snake-case-fix

luc122c commented:
@danielroe Some of these test cases may not be possible with the current logic. For example, `TESTAppName` I can't figure out a way for this to work, since TypeScript doesn't know that 'Test', 'App' and 'Name' are words, so it doesn't know to split 'TESTA' before the A.

danielroe committed:
Merge remote-tracking branch 'origin/main' into upper-snake-case-fix

danielroe committed:
refactor: use type from `scule`

danielroe commented:
The logic isn't related to whether they are words. Note that scule splits these all correctly; it's based on splitting before the change of case.

I think we can use scule's native types for this - but we should ideally expose them upstream.

luc122c commented:
> The logic isn't related to whether they are words. Note that scule splits these all correctly; it's based on splitting before the change of case.

[This is the closest I've managed to get](https://www.typescriptlang.org/play?ts=5.2.2#code/C4TwDgpgBAKg9gVTJATgZQHYEMDWEDCWAzhADwxQQAewEGAJkVEcCgJYYDmANFAAooIANzZwArk2q0GTFuy5QAvFADkKgHxLYlGnUZQABgBIA3hwBmEFFHxiUgjMAC+pi1agAlCCycGAUFBQAPyGpgAUAsKiEjrS+gAycADuVgDGxGSRIuJE6gGBwVBhtvZ0wLF6TEio6SSkJQ7AmiERgtkxUpWqAPoqhWpQAFytUTkVMqoqQWrDDWXj+iq90yqDSyoAlFtDkxs7ahsuJtVpGfV2jepH8Cfo2HiEdV4svHOOV-6Bg7AA3H5+oEgsG8wAAjFobsgrJhcAQzipaCwAILIAByWAAthAND8oH4APT4gpQAB6QQB4GgMBBACYIYgoXdYY8yCoYABRNAwbpIvh8bqopEAWXZOLxhOJZIpQOpLAAzPTbjCHvDEcAeXyBcLRepcQSiQUpYCqSCACyKxnKuF1BEg7pYZBakVi-WS8nG4EsACsFtQVpZpFtyLR2pdEsN7spnuAADZfdD7tbWWreXx0ViwwbAkao7LgAB2eNMlU2lOa9PY3XirOkyMykEADiL-tVdtTAsxlb14ezdZNLAAnM3EwG2ZyYCiwBXM26gA) separately, just following the logic.

> I think we can use scule's native types for this - but we should ideally expose them upstream.

Does Scule expose itself as a type, [like Defu does](https://github.com/unjs/defu?tab=readme-ov-file#type)?

danielroe commented:
No - see https://github.com/unjs/scule/pull/58. I pushed a change in [`c5f9a49` (#23696)](https://github.com/nuxt/nuxt/pull/23696/commits/c5f9a491ea6a622e77f3d2ab36ebe96df4d04604) that uses the inferred return type of `snakeCase`, however.

luc122c committed:
Merge commit '5fea1728830777ac8942c74ba8751822716a7aae' into upper-snake-case-fix

luc122c committed:
test: use SnakeCase type rather than returntype of snakeCase function

danielroe commented on packages/schema/src/types/config.ts:
```diff
@@ -4,50 +4,16 @@ import type { Options as VuePluginOptions } from '@vitejs/plugin-vue'
 import type { Options as VueJsxPluginOptions } from '@vitejs/plugin-vue-jsx'
 import type { SchemaDefinition } from 'untyped'
 import type { NitroRuntimeConfig, NitroRuntimeConfigApp } from 'nitropack'
+// TODO: expose SnakeCase type from scule
```
```suggestion
```

luc122c replied:
@danielroe Would https://github.com/unjs/scule/pull/58 need to be merged and released first?

danielroe replied:
Yes, it will.

danielroe commented on test/fixtures/basic-types/types.ts:
```diff
@@ -271,6 +271,23 @@ describe('runtimeConfig', () => {
   })
+
+  it('correctly converts different kinds of names to snake case', () => {
+    // https://github.com/unjs/nitro/blob/be25a21d79acf272b3629d69332d66795ae90b6e/src/runtime/config.ts#L56
```
```suggestion
```

danielroe commented on test/fixtures/basic-types/types.ts:
```diff
@@ -271,6 +271,23 @@ describe('runtimeConfig', () => {
+
+  it('correctly converts different kinds of names to snake case', () => {
+    // https://github.com/unjs/nitro/blob/be25a21d79acf272b3629d69332d66795ae90b6e/src/runtime/config.ts#L56
+    // require('scule').snakeCase(key).toUpperCase()
```
```suggestion
```

luc122c committed:
Apply suggestions from code review

Co-authored-by: Daniel Roe <daniel@roe.dev>

luc122c committed:
Merge commit '37d64a615b7bb4a13a2e722a925764100f832af3' into upper-snake-case-fix

luc122c committed:
chore(deps): upgrade scule

luc122c committed:
chore: remove comment

scule 1.1.0 has now been released.
https://github.com/unjs/scule/releases/tag/v1.1.0

Co-authored-by: Daniel Roe <daniel@roe.dev>

luc122c committed:
Merge branch 'main' into upper-snake-case-fix
